Added custom Log Formatter

memije opened the pull request:
Added custom log formatter for gin logging middleware.

This formatter will create a log entry in JSON format with the following information:

- Request's timestamp in RFC3339 format
- Client's IP address
- Consumed API path/endpoint
- Request's response status code

It will also include the following fields if a `CognitoCache` instance
is in the gin's context:

- Username
- User's groups

---

Simple usage example:

```go
package main

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/protsack-stephan/gin-toolkit/httpmw"
)

func main() {
	g := gin.New()
	g.Use(gin.Recovery())
	g.Use(gin.LoggerWithFormatter(httpmw.LogFormatter))

	g.GET("/", func(c *gin.Context) {
		c.Status(http.StatusTeapot)
	})

	g.Run()
}
```

Produced output:

```js
{"request_time":"2022-05-05T12:46:21-05:00","ip":"::1","path":"/","status":418,"username":"","user_groups":null}
```

Note `username` and `user_groups` are empty due to missing `CognitoCache` user in gin's context

memije committed:
Added custom Log Formatter

Added custom log formatter for gin logging middleware.
This formatter will create a log entry in JSON format with the
following information:

- Request's timestamp in RFC3339 format
- Client's IP address
- Consumed API path/endpoint
- Request's response status code

It will also include the following fields if a `CognitoCache` instance
is in the gin's context:

- Username
- User's groups

protsack-stephan commented on httpmw/logging_test.go:
```diff
@@ -0,0 +1,92 @@
+			e := gin.New()
+			out := new(bytes.Buffer)
+			apiPath := "/test"
+			testIP := "0.0.0.0"
```
Those looks like they can be `const`(s) or at least defined in outside test cases in the main function.

protsack-stephan commented on httpmw/logging_test.go:
```diff
@@ -0,0 +1,92 @@
+			assert.Equal(http.StatusTeapot, w.Code)
+
+			entry := new(logEntry)
+			err = json.Unmarshal(out.Bytes(), entry)
```
This assignment is redundant we can just do `assert.NoError(json.Unmarshal(out.Bytes(), entry))`.

protsack-stephan commented on httpmw/logging_test.go:
```diff
@@ -0,0 +1,92 @@
+		assert.Equal(http.StatusTeapot, w.Code)
+
+		entry := new(logEntry)
+		err = json.Unmarshal(out.Bytes(), entry)
```
This assignment is redundant we can just do `assert.NoError(json.Unmarshal(out.Bytes(), entry))`.

protsack-stephan reviewed (commented):
LGTM! Couple small comments.

memije committed:
Fixed comments on logging test

- Added `apiPath` and `testIP` as constants
- Removed redundant error assignment when doing json unmarshaling

prabhat393 commented:
Looks good.
